examples/energy-management-app/energy-management-common/common/src/EnergyManagementAppCommonMain.cpp

```diff
@@ -0,0 +1,410 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "EnergyManagementAppCommonMain.h"
+#include "EnergyManagementAppCmdLineOptions.h"
+#include <DeviceEnergyManagementDelegateImpl.h>
+#include <DeviceEnergyManagementManager.h>
+#include <ElectricalPowerMeasurementDelegate.h>
+#include <EnergyEvseMain.h>
+#include <PowerTopologyDelegate.h>
+#include <WhmMain.h>
+#include <WhmManufacturer.h>
+#include <app/clusters/electrical-energy-measurement-server/electrical-energy-measurement-server.h>
+#include <device-energy-management-modes.h>
+#include <energy-evse-modes.h>
+#include <water-heater-mode.h>
+
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/data-model/Nullable.h>
+#include <lib/support/CodeUtils.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::DataModel;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::DeviceEnergyManagement;
+using namespace chip::app::Clusters::ElectricalPowerMeasurement;
+using namespace chip::app::Clusters::ElectricalEnergyMeasurement;
+using namespace chip::app::Clusters::PowerTopology;
+using namespace chip::app::Clusters::WaterHeaterManagement;
+
+namespace {
+
+std::unique_ptr<DeviceEnergyManagementDelegate> gDEMDelegate;
+std::unique_ptr<DeviceEnergyManagementManager> gDEMInstance;
+std::unique_ptr<ElectricalPowerMeasurementDelegate> gEPMDelegate;
+std::unique_ptr<ElectricalPowerMeasurementInstance> gEPMInstance;
+std::unique_ptr<PowerTopologyDelegate> gPTDelegate;
+std::unique_ptr<PowerTopologyInstance> gPTInstance;
+// Electrical Energy Measurement cluster uses ember to initialise
+std::unique_ptr<ElectricalEnergyMeasurementAttrAccess> gEEMAttrAccess;
+bool gCommonClustersInitialized = false;
+
+/*
+ *  @brief  Creates a Delegate and Instance for PowerTopology clusters
+ *
+ * The Instance is a container around the Delegate, so
+ * create the Delegate first, then wrap it in the Instance
+ * Then call the Instance->Init() to register the attribute and command handlers
+ */
+CHIP_ERROR PowerTopologyInit(chip::EndpointId endpointId)
```

Comment:
I was (and may suggest that we still do) move this to a new EnergyReportingMain.cpp so that PowerTopology/EEM/EPM are moved into a re-usable component. That's a change I have in my 'refactoring' PR - but let's get this PR merged.

I think I had a DEMMain as well so that the DEM related code isn't necessarily in here.

Reply:
Agreed, that's a good idea. Let's do that in a follow up smaller PR.